providers/azure.go

```diff
@@ -103,18 +103,18 @@ func overrideTenantURL(current, defaultURL *url.URL, tenant, path string) {
 	}
 }
 
-func (p *AzureProvider) GetLoginURL(redirectURI, state, _ string) string {
+func (p *AzureProvider) GetLoginURL(redirectURI, state, _, codeChallenge, codeChallengeMethod string) string {
```

Comment:
Are all these providers ready to support this? For ones that aren't, we should name the arguments `_` and not use these even if a user sets the flag I think?

Reply:
I can't speak for most providers, but many are. Azure, Google, Dex, Auth0, Okta,.. do support PKCE today.

- I see your point about setting this to `_` for some providers that do not support it but I can't test _all_ providers to be sure. I also didn't want to limit this to feature to the ones only we know of supporting it
- As stated in the commit message and earlier comments, I do a few things here to try and make this introduction of PKCE "safe"
  - oauth2-proxy first attempts to detect if a server supports PKCE by looking at the discovery document for supported methods. If a provider explicitly defines code challenge methods, we enable PKCE without user interaction
  - If oauth2-proxy cannot find the provider advertising support for PKCE, we turn it off.
  - Some providers support PKCE but do not yet implement the code challenge advertisment in their discovery document (looking at you Azure), so a force flag is introduced to allow the user to enable it knowing their provider does support PKCE

If we went with what you said, we'd need to release a whole new version of oauth2-proxy as providers add support for PKCE, I think how I implemented it here is more realistic and we will be able to support PKCE as providers update to this standard.

Note: If a provider doesn't support PKCE and oauth2-proxy attempts it anyways, the OAuth2 spec specifies that authorization servers should ignore extra URL parameters. So _ideally_ nothing would change. However, in the wild it was found that some providers do have errors if PKCE is presented when it doesn't expect it (LinkedIn)

Reply:
> Note: If a provider doesn't support PKCE and oauth2-proxy attempts it anyways, the OAuth2 spec specifies that authorization servers should ignore extra URL parameters. So ideally nothing would change. However, in the wild it was found that some providers do have errors if PKCE is presented when it doesn't expect it (LinkedIn)

This is the one area I'm worried about (so maybe we just need to advertize this as a potential breaking change in teh CHANGELOG).

Specifically the ecosystem under the OIDC provider (likely our most popular) -- there are SOOOOO many and they are are special snowflakes that decide the break the spec in weird ways (our PR/issue history is littered with people adding corner case support for their OIDC IdP's werid implementation of the spec).

This is the one area I'm potentially concerned about opting all providers in by default (so either we have this as an opt-in functionality via flag, or if its on by default, we need to support an opt-out)?

I didn't fully follow the flag logic closely -- is this on by default now & the force flag forces PKCE for OIDC's who's discovery documents don't mention it?

Reply:
Ah, I see you touch on Okta's issues with this here: https://github.com/oauth2-proxy/oauth2-proxy/pull/1541#discussion_r809581470

I have a strong suspicion if even Okta has issues in supporting, a lot of the fringe OIDC IdPs we've had bug/issue reports about over the years will have issues with this.